monkey/common/di_container.py

```diff
@@ -56,12 +61,45 @@ def register_instance(self, interface: Type[T], instance: T):
         self._instance_registry[interface] = instance
         DIContainer._del_key(self._type_registry, interface)
 
+    def register_convention(self, type_: Type[T], name: str, instance: T):
+        """
+        Register an instance as a convention
+
+        At times — particularly when dealing with primative types — it can be useful to define a
+        convention for how dependencies should be resolved. For example, you might want any class
+        that specifies `hostname: str` in its constructor to receive the hostname of the system it's
+        running on. Registering a convention allows you to assign an object instance to a type, name
+        pair.
+
+        Example:
+            class TestClass:
+                def __init__(self, hostname: str):
+                    self.hostname = hostname
+
+            di_container = DIContainer()
+            di_container.register_convention(str, "hostname", "my_hostname.domain")
```

Comment:
What happens if I have a class `ConnectionChecker(scanner: INetworkScanner, endpoint: str)`. Is it possible to create and inject the scanner, but pass an endpoint dynamically (we don't know what endpoints we'll want to check at build time)? What if someone has already defined an endpoint convention?

Reply:
I guess objects at DI container need to be created in their entirety or not at all. So `scanner` would be created and passed down the call stack until it would actually get used.